### PR TITLE
Fix implementation of BYTE

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/formal/byte_big.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+rule = Rule()
+
+n_bits = 256
+
+# Input vars
+X = BitVec('X', n_bits)
+A = BitVec('A', n_bits)
+
+rule.require(A >= n_bits / 8)
+rule.check(BYTE(A, X), 0)
```
Comment (chriseth) [inline test/formal/byte_big.py:18]: The old implementation of BYTE finds a counterexample here

File: test/formal/opcodes.py [hunk 2]
```diff
@@ -58,5 +58,10 @@ def SAR(x, y):
 	return y >> x
 
 def BYTE(i, x):
+	bitInt = (BV2Int(i) + 1) * 8
 	bit = (i + 1) * 8
-	return If(UGT(bit, x.size()), BitVecVal(0, x.size()), (LShR(x, (x.size() - bit))) & 0xff)
+	return If(
+		bitInt > x.size(),
```
Comment (ekpyron) [inline test/formal/opcodes.py]: Would
```suggestion
		i > x.size() / 8 - 1,
```
also work?

Comment (ekpyron) [reply test/formal/opcodes.py]: (I guess we can assume ``x.size()`` to be a multiple of 8 and greater than 1 and then it's the same without any casting or non-constant arithmetic?)

**Comment by chriseth** [comment]
Rebased.

**Review by ekpyron — DISMISSED** [review]

File: test/formal/opcodes.py [hunk 3]
```diff
@@ -59,4 +59,8 @@ def SAR(x, y):
 
 def BYTE(i, x):
 	bit = (i + 1) * 8
-	return If(UGT(bit, x.size()), BitVecVal(0, x.size()), (LShR(x, (x.size() - bit))) & 0xff)
+	return If(
+		i > x.size() / 8 - 1,
```
Comment (hrkrshnn) [inline test/formal/opcodes.py]: Isn't this a signed comparison?

For example:

```py
from z3 import *

def fun(i, x):
    return i > x.size() / 8 - 1

x = BitVecVal(200, 32)
i = BitVecVal(2**32 - 1, 32)

simplify(fun(i, x))
# returns False
```

Comment (chriseth) [reply test/formal/opcodes.py]: Fixed

**Review by ekpyron — APPROVED** [review]

**Pull request merged**